cmake: linker_script: Fix using deferred init with `CMAKE_LINKER_GENERATOR`

pdgendt opened the pull request:
The section for deferred devices was missing for `CMAKE_LINKER_GENERATOR`.

Issue shown here:
```shell
$ west build -p -b qemu_cortex_m3 tests/kernel/device -- -DCONFIG_CMAKE_LINKER_GENERATOR=y
# ...
/opt/zephyr-sdk-0.16.5-1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: zephyr/kernel/libkernel.a(init.c.obj): in function `z_impl_device_init':
/home/pdgendt/dev/zephyr-main/zephyr/kernel/init.c:388: undefined reference to `__deferred_init_list_start'
/opt/zephyr-sdk-0.16.5-1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/12.2.0/../../../../arm-zephyr-eabi/bin/ld.bfd: /home/pdgendt/dev/zephyr-main/zephyr/kernel/init.c:388: undefined reference to `__deferred_init_list_end'
```

Regression introduced by #67335

tejlmand commented on cmake/linker_script/common/common-rom.cmake:
```diff
@@ -7,6 +7,7 @@ zephyr_linker_section_obj_level(SECTION init LEVEL PRE_KERNEL_2)
 zephyr_linker_section_obj_level(SECTION init LEVEL POST_KERNEL)
 zephyr_linker_section_obj_level(SECTION init LEVEL APPLICATION)
 zephyr_linker_section_obj_level(SECTION init LEVEL SMP)
+zephyr_iterable_section(NAME _deferred_init KVMA RAM_REGION GROUP RODATA_REGION SUBALIGN 4)
```
why the `_` prefix in this case ?
We don't use that elsewhere for the linker generator.

symbols are automatically prefixed with `__` in places like this for referencing in consistent ways:
https://github.com/zephyrproject-rtos/zephyr/blob/d3c9a986ec30e4c1fde5092e73d76f7020ea20ee/cmake/linker/ld/ld_script.cmake#L220

pdgendt replied:
`zephyr_iterable_section` creates `_<name>_list_start` and `_<name>_list_end` and those were manually added for deferred devices:

https://github.com/zephyrproject-rtos/zephyr/blob/d3c9a986ec30e4c1fde5092e73d76f7020ea20ee/include/zephyr/linker/common-rom/common-rom-kernel-devices.ld#L21-L23

pdgendt replied:
Hmn, the `z_deferred_init` might not match now..

pdgendt committed:
tests: kernel: device: Add CMAKE_LINKER_GENERATOR test case

Add a test case entry with CONFIG_CMAKE_LINKER_GENERATOR enabled.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
cmake: linker_script: Support deferred init with CMAKE_LINKER_GENERATOR

The section for deferred devices was missing when
CONFIG_CMAKE_LINKER_GENERATOR is enabled.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
Rebased due to merge conflict

cfriedt reviewed (approved):
LGTM
